src/bazel/bazel_quickpick.ts

```diff
@@ -169,10 +173,108 @@ export async function queryQuickPickPackage({
   const packagePaths = await new BazelQuery(
     getDefaultBazelExecutablePath(),
     workspaceInfo.workspaceFolder.uri.fsPath,
-  ).queryPackages(query ?? "//...");
+  ).queryPackages(query ?? "//...", { abortSignal });
 
   // Sort the labels so the QuickPick is ordered.
   return packagePaths
     .sort()
     .map((target) => new BazelTargetQuickPick("//" + target, workspaceInfo));
 }
+
+/**
+ * Shows a QuickPick of Bazel labels that dynamically updates its items based on the user's input.
+ *
+ * @param options Configuration options for the QuickPick
+ * @returns A promise that resolves with the selected BazelTargetQuickPick, or undefined if no selection was made
+ */
+export function showDynamicQuickPick(options: {
+  query?: string;
+  queryFunctor: (params: QuickPickParams) => Promise<BazelTargetQuickPick[]>;
+  workspaceInfo?: BazelWorkspaceInfo;
+}): Promise<BazelTargetQuickPick | undefined> {
+  const quickPick = vscode.window.createQuickPick<BazelTargetQuickPick>();
+  quickPick.title = "Select a Bazel target";
+  quickPick.placeholder = "Start typing to search for targets...";
+  quickPick.matchOnDescription = true;
+  quickPick.matchOnDetail = true;
+
+  let abortController: AbortController | undefined = new AbortController();
+  let timeout: NodeJS.Timeout | undefined;
+
+  const updateQuickOptions = async (value: string): Promise<void> => {
+    // Cancel any previous query
+    abortController?.abort();
+
+    // Store the current abort controller to ensure we don't cancel a new query
+    const currentAbortController = new AbortController();
+    abortController = currentAbortController;
+
+    // Show loading state
+    quickPick.busy = true;
+    quickPick.items = [];
+
+    // Process the input: keep everything before the last slash and add '...'
+    const newQuery = value.includes("/")
+      ? options.query.replace(
+          "//...",
+          `${value.substring(0, value.lastIndexOf("/"))}/...`,
+        )
+      : options.query;
+
+    try {
+      const items = await options.queryFunctor({
+        query: newQuery,
+        workspaceInfo: options.workspaceInfo,
+        abortSignal: currentAbortController.signal,
+      });
+      quickPick.items = items;
+    } catch (error) {
+      if (error instanceof Error && error.message === "Query was aborted") {
```

Comment:
Checking the message here seems slightly brittle. If we instead rejected with the AbortSignal `reason` (which by default is a AbortError DOMException), then we could check for this, e.g. by using `error.name == "AbortError"`.